File: bindings/java/src/main/java/dev/invoicekit/EngineResult.java
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 The InvoiceKit Authors

package dev.invoicekit;

import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.Objects;

/** Copied Engine ABI result bytes plus the C ABI status code. */
public final class EngineResult {
    /** C ABI status for successful Engine ABI responses. */
    public static final int STATUS_OK = 0;
    /** C ABI status for canonical Engine ABI error responses. */
    public static final int STATUS_ERROR = 1;
    /** C ABI status for invalid native handles. */
    public static final int STATUS_INVALID_HANDLE = 2;

    private final int statusCode;
    private final byte[] responseBytes;

    /**
     * Create a result from copied response bytes.
     *
     * @param statusCode C ABI status code.
     * @param responseBytes response bytes copied from native memory or the REST sidecar.
     */
    public EngineResult(int statusCode, byte[] responseBytes) {
        this.statusCode = statusCode;
        this.responseBytes = Objects.requireNonNull(responseBytes, "responseBytes").clone();
    }

    /**
     * Return the C ABI status code.
     *
     * @return numeric status code reported by the backing runtime.
     */
    public int statusCode() {
        return statusCode;
    }

    /**
     * Return true when the backing runtime reported a successful Engine ABI response.
     *
     * @return true when {@link #statusCode()} is {@link #STATUS_OK}.
     */
    public boolean isOk() {
        return statusCode == STATUS_OK;
    }

    /**
     * Return a defensive copy of the response bytes.
     *
     * @return copied response bytes.
     */
    public byte[] responseBytes() {
        return responseBytes.clone();
    }

    /**
     * Decode response bytes as UTF-8 text.
     *
     * @return response bytes decoded as UTF-8.
     */
    public String responseText() {
        return new String(responseBytes, StandardCharsets.UTF_8);
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) {
            return true;
        }
        if (!(other instanceof EngineResult that)) {
            return false;
        }
        return statusCode == that.statusCode && Arrays.equals(responseBytes, that.responseBytes);
    }

    @Override
    public int hashCode() {
        return 31 * Integer.hashCode(statusCode) + Arrays.hashCode(responseBytes);
    }

    @Override
    public String toString() {
        return "EngineResult{statusCode=" + statusCode + ", responseBytes=" + responseBytes.length + " bytes}";
    }
}

File: bindings/dotnet/src/InvoiceKit/EngineResult.cs
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 The InvoiceKit Authors

using System.Text;

namespace InvoiceKit;

/// <summary>
/// Copied Engine ABI response bytes plus the C ABI status code.
/// </summary>
public sealed class EngineResult : IEquatable<EngineResult>
{
    private readonly byte[] responseBytes;

    /// <summary>
    /// Create an Engine ABI result from copied response bytes.
    /// </summary>
    /// <param name="statusCode">Status code returned by the C ABI or REST sidecar.</param>
    /// <param name="responseBytes">Response bytes copied from native memory or HTTP response content.</param>
    public EngineResult(EngineStatusCode statusCode, byte[] responseBytes)
    {
        StatusCode = statusCode;
        this.responseBytes = (responseBytes ?? throw new ArgumentNullException(nameof(responseBytes))).ToArray();
    }

    /// <summary>
    /// Status code returned by the C ABI or REST sidecar.
    /// </summary>
    public EngineStatusCode StatusCode { get; }

    /// <summary>
    /// True when <see cref="StatusCode"/> is <see cref="EngineStatusCode.Ok"/>.
    /// </summary>
    public bool IsOk => StatusCode == EngineStatusCode.Ok;

    /// <summary>
    /// Return a defensive copy of the Engine ABI response bytes.
    /// </summary>
    /// <returns>Copied response bytes.</returns>
    public byte[] ResponseBytes() => responseBytes.ToArray();

    /// <summary>
    /// Decode response bytes as UTF-8 text.
    /// </summary>
    /// <returns>Response bytes decoded as UTF-8.</returns>
    public string ResponseText() => Encoding.UTF8.GetString(responseBytes);

    /// <inheritdoc />
    public bool Equals(EngineResult? other)
    {
        return other is not null
            && StatusCode == other.StatusCode
            && responseBytes.SequenceEqual(other.responseBytes);
    }

    /// <inheritdoc />
    public override bool Equals(object? obj) => Equals(obj as EngineResult);

    /// <inheritdoc />
    public override int GetHashCode()
    {
        var hash = new HashCode();
        hash.Add(StatusCode);
        foreach (var b in responseBytes)
        {
            hash.Add(b);
        }

        return hash.ToHashCode();
    }

    /// <inheritdoc />
    public override string ToString() => $"EngineResult {{ StatusCode = {StatusCode}, ResponseBytes = {responseBytes.Length} bytes }}";
}
